doc/kernel/drivers/index.rst

```diff
@@ -122,36 +125,42 @@ Most drivers will be implementing a device-independent subsystem API.
 Applications can simply program to that generic API, and application
 code is not specific to any particular driver implementation.
 
+If all driver API instances are assigned to their respective API linker section
+use :c:macro:`DEVICE_API_IS()` to verify the API's type.
+
 A subsystem API definition typically looks like this:
 
 .. code-block:: C
 
   typedef int (*subsystem_do_this_t)(const struct device *dev, int foo, int bar);
   typedef void (*subsystem_do_that_t)(const struct device *dev, void *baz);
 
-  struct subsystem_api {
+  __subsystem struct subsystem_driver_api {
         subsystem_do_this_t do_this;
         subsystem_do_that_t do_that;
   };
 
   static inline int subsystem_do_this(const struct device *dev, int foo, int bar)
   {
-        struct subsystem_api *api;
+        CHECKIF(!DEVICE_API_IS(subsystem, dev)) {
```

Comment:
This implies we will never load a driver at runtime, and that we're willing to accept runtime checks forever.

Truly, this would be better service if the first parameter instead were properly typed.

Reply:
I'm not against this in any way, but this would be a huge endeavor to change from the current implementation.

- For runtime loadable drivers; we're still an RTOS, I mean this sounds nice in theory, but how do you envision this? Currently the hardware in DTS is the single source of truth for firmware, macros expand to constants.
- Downstream device drivers would be broken, and would need a larger migration path
- Who will take up this task/responsibility?

Reply:
I think eventually we should move to something like:

```c
struct gpio_device {
    struct device dev;
    ...
};
```

What we have now is a bad/poor design that forces us to do even more hacks to get functionality that could have been offered by the compiler itself from day one (not the first case in Zephyr). I'm not sure the Zephyr community is psychologically prepared for another massive breaking change, but if we expect to have a proper device model, I don't think there'll be a smooth path or any other way. By proper, I mean, roughly:

- Devices with strong typing (aka `struct gpio_device`).
- Better integration with DT (maybe a spec that allows to specify what to dump from DT to device's config in ROM). Existing macrology is just insane.
- Better decoupling with DT, so a device can be instantiated from anywhere (related with above: DT would only mandate a source of device's config and created instances)
- Better device dependencies (no ELF hackery), including runtime ones
- Better init sequence, de-init for runtime reconfiguration of IP blocks that can be multifunctional (timers, serial...)
- Limit HAL usage: only stateless HALs, or offer something like svd2rust/chiptool for consistent device driver code
- etc etc

I won't block this PR but neither approve, I leave it to the community to decide here. But IMHO, while this PR may solve some immediate problems, it is also adding more technical debt.

Reply:
As soon as this is taken as a temporary workaround for our weak device model, I'm fine with this. Consider reassigning if I become a blocker.